### PR TITLE
IGNITE-13796 docs: update docs and examples for kubernetes module

### PR DESCRIPTION
**Pull request by timoninmaxim**

### CODE REVIEW

File: docs/_docs/code-snippets/k8s/stateful/node-configuration.xml [hunk 1]
```diff
@@ -16,11 +16,17 @@
...
 -->
 <beans xmlns="http://www.springframework.org/schema/beans"
-    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
-    xsi:schemaLocation="http://www.springframework.org/schema/beans         
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://www.springframework.org/schema/beans
     http://www.springframework.org/schema/beans/spring-beans.xsd">
 
     <!-- tag::config-block[] -->
+    <bean class="org.apache.ignite.kubernetes.configuration.KubernetesConnectionConfiguration" name="k8sCfg">
```
Comment (dmagda) [inline docs/_docs/code-snippets/k8s/stateful/node-configuration.xml:24]: I would put this bean definition into the TcpDiscoveryKubernetesIpFinder bean right away. Simplifies readability. Plus, the KubernetesConnectionConfiguration bean is used only by the TcpDiscoveryKubernetesIpFinder bean; thus it should be fine to put one inside of another right away.

Please consider such a change for similar places below.

File: docs/_docs/installation/kubernetes/generic-configuration.adoc [hunk 2]
```diff
@@ -388,6 +388,17 @@ Note that we use the external IP address (LoadBalancer Ingress) of the service.
 include::{javaFile}[tags=connectThinClient, indent=0]
 ----
 
+Thin clients support the partition awareness feature that may speed up operations. Thin clients store a mapping of Ignite nodes to cache partitions and send partition requests directly to the right node.
+A client connected to the Kubernetes cluster through Ingress can't build such a map. Instead, to activate the feature, one should start a client within the cluster and configure it with KubernetesConnectionConfiguration.
```
Comment (dmagda) [inline docs/_docs/installation/kubernetes/generic-configuration.adoc:392]: A not for Nikita Safonov, who will be reviewing and editing the doc. 

I would probably put these details on the thin client documentation page. And add a callout to this Kubernetes pages saying that you need to enable a thin-client specific IP finder (providing a reference to the thin client page for mode details).

File: docs/_docs/code-snippets/k8s/stateful/node-configuration.xml (hunk 1 above)
Comment (timoninmaxim) [reply docs/_docs/code-snippets/k8s/stateful/node-configuration.xml]: fixed

File: docs/_docs/installation/kubernetes/generic-configuration.adoc (hunk 2 above)
Comment (timoninmaxim) [reply docs/_docs/installation/kubernetes/generic-configuration.adoc:392]: Hi, I'm agree that there should be part of ClientAddressFinder in thin client. As any user can write own implementation of this interface and use it within any (not only k8s env).

@dmagda @Nikita-tech-writer Nikita, if you didn't start work on this ticket yet I will add this to docs. Please let me know you current status on this ticket

Comment (Nikita-tech-writer) [reply docs/_docs/installation/kubernetes/generic-configuration.adoc:392]: Hi @timoninmaxim,
Yes, I will start to work on this ticket after the NY holidays, so you can add it to docs for sure. Thank you.

Comment (timoninmaxim) [reply docs/_docs/installation/kubernetes/generic-configuration.adoc]: @Nikita-tech-writer hi! I've updated docs and push. Please let me know if it is ok or not.

Comment (Nikita-tech-writer) [reply docs/_docs/installation/kubernetes/generic-configuration.adoc]: Hi @timoninmaxim , I'll give you the feedback shortly.

File: docs/_docs/installation/kubernetes/generic-configuration.adoc [hunk 3]
```diff
@@ -388,6 +388,20 @@ Note that we use the external IP address (LoadBalancer Ingress) of the service.
 include::{javaFile}[tags=connectThinClient, indent=0]
 ----
 
+== Partition Awareness
+
+include::../../includes/partition-awareness.adoc[]
+
+To enable the partition awareness feature within scaling Kubernetes enviroment one should start a client within the cluster and configure it with `KubernetesConnectionConfiguration`.
```
Comment (Nikita-tech-writer) [inline docs/_docs/installation/kubernetes/generic-configuration.adoc:395]:
```suggestion
To enable the partition awareness feature within scaling Kubernetes enviroment, one should start a client within the cluster and configure it with `KubernetesConnectionConfiguration`.
```

File: docs/_docs/installation/kubernetes/generic-configuration.adoc [hunk 4]
```diff
@@ -388,6 +388,20 @@ Note that we use the external IP address (LoadBalancer Ingress) of the service.
...
+In this case, a client can connect to every pod in a cluster.
+
+Note that client configuration must be in sync with Ignite node configuration.
```
Comment (Nikita-tech-writer) [inline docs/_docs/installation/kubernetes/generic-configuration.adoc:398]:
```suggestion
Note that client configuration must be in sync with the Ignite node configuration.
```

File: docs/_docs/thin-clients/java-thin-client.adoc [hunk 5]
```diff
@@ -86,6 +86,16 @@ The following code sample illustrates how to use the partition awareness feature
 include::{sourceCodeFile}[tag=partition-awareness,indent=0]
 ----
 
+If list of server nodes is dynamically changing or scaling than it is possible to configure connection with custom implementation of `ClientAddressFinder`. It should provide actual server addresses collection for every time client ask it.
```
Comment (Nikita-tech-writer) [inline docs/_docs/thin-clients/java-thin-client.adoc:89]:
```suggestion
If a list of server nodes is dynamically changing or scaling, then it is possible to configure the connection with custom implementation of `ClientAddressFinder`. It should provide a number of current server addresses every time a client asks for them.
```

File: docs/_docs/thin-clients/java-thin-client.adoc [hunk 6]
```diff
@@ -86,6 +86,16 @@ The following code sample illustrates how to use the partition awareness feature
...
+The following code sample illustrates how to use it.
+
+[source, java]
+----
+include::{sourceCodeFile}[tag=client-address-finder,indent=0]
+----
+
+Currently Ignite provides the only implementation of this interface. `ThinClientKubernetesAddressFinder` is created to handle scalable Kubernetes environment.
```
Comment (Nikita-tech-writer) [inline docs/_docs/thin-clients/java-thin-client.adoc:97]:
```suggestion
Currently, Ignite provides the only implementation of this interface. `ThinClientKubernetesAddressFinder` is created to handle scalable Kubernetes environment.
```

**Review by Nikita-tech-writer — COMMENTED** [review]
Please see my minor suggested changes to this PR.

**Comment by Nikita-tech-writer** [comment]
@dmagda Could you please also take a look?

File: docs/_docs/installation/kubernetes/generic-configuration.adoc (hunk 4 above)
Comment (dmagda) [inline docs/_docs/installation/kubernetes/generic-configuration.adoc:398]: What does this sentence mean? What type of synchronization is assumed? Confusing.

File: docs/_docs/thin-clients/java-thin-client.adoc (hunk 6 above)
Comment (dmagda) [reply docs/_docs/thin-clients/java-thin-client.adoc:97]: I don't understand what the sentence means. You say "Ignite provides the only implementation of this interface". What interface is assumed? Why is it important to mention this only implementation?

Instead, provide a bullet-point list explaining the code snippet. The snippet has the `fetchServerAddresses` method that is not implemented. It seems that you are trying to say that "The fetchServerAddress() method shows how an example implementation might look like if you want clients to retrieve the server nodes' addresses dynamically. If you are running in a Kubernetes environment then use the `ThinClientKubernetesAddressFinder` in your thin client configuration to enable the automatic resolution of the server addresses on the client side".  Nikita please help to adjust the latter if Maxim confirms that my understanding is correct.

File: examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.ignite.Ignition;
+import org.apache.ignite.client.ClientCache;
+import org.apache.ignite.client.IgniteClient;
+import org.apache.ignite.client.ThinClientKubernetesAddressFinder;
+import org.apache.ignite.configuration.ClientConfiguration;
+import org.apache.ignite.examples.model.Address;
+import org.apache.ignite.kubernetes.configuration.KubernetesConnectionConfiguration;
+
+/**
+ * Demonstrates how to use Ignite thin client within the Kubernetes cluster using KubernetesConnectionConfiguration.
```
Comment (dmagda) [inline examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java:29]: @Nikita-tech-writer, please check the comments in the source code files too. I see some spellings below.

Comment (Nikita-tech-writer) [reply examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java:29]: @dmagda OK, sure

File: docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/JavaThinClient.java [hunk 8]
```diff
@@ -360,6 +360,26 @@ void partitionAwareness() throws Exception {
...
+            return dynamicServerAddresses;
+        };
+
+        ClientConfiguration cfg = new ClientConfiguration()
+            .setAddressFinder(finder)
+            .setPartitionAwarenessEnabled(true);
+
+        try (IgniteClient client = Ignition.startClient(cfg)) {
+            ClientCache<Integer, String> cache = client.cache("myCache");
+            // Put, get or remove data from the cache...
```
Comment (Nikita-tech-writer) [inline docs/_docs/code-snippets/java/src/main/java/org/apache/ignite/snippets/JavaThinClient.java:377]:
```suggestion
            // Put, get, or remove data from the cache...
```

File: examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+ * <p>
+ * Prerequisites:
+ * <ul>
+ * <li>Running Ignite Kubernetes cluster. Check modules/kubernetes/DEVNOTES.md for example how to run a cluster locally.</li>
+ * <li>A thin client application should be run within the cluster to have access to Ignite nodes pods.</li>
```
Comment (Nikita-tech-writer) [inline examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java:34]:
```suggestion
 * <li>A thin client application should be run within the cluster to provide access to the Ignite nodes pods.</li>
```

File: examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.ignite.configuration.ClientConfiguration;
+import org.apache.ignite.examples.model.Address;
+import org.apache.ignite.kubernetes.configuration.KubernetesConnectionConfiguration;
+
+/**
+ * Demonstrates how to use Ignite thin client within the Kubernetes cluster using KubernetesConnectionConfiguration.
+ * <p>
+ * Prerequisites:
+ * <ul>
+ * <li>Running Ignite Kubernetes cluster. Check modules/kubernetes/DEVNOTES.md for example how to run a cluster locally.</li>
```
Comment (Nikita-tech-writer) [inline examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java:33]:
```suggestion
 * <li>Running Ignite Kubernetes cluster. Check modules/kubernetes/DEVNOTES.md as an example of how to run a cluster locally.</li>
```

File: modules/kubernetes/DEVNOTES.md [hunk 11]
```diff
@@ -34,7 +34,7 @@ Create a folder for all the files to be placed in the Docker image and copy the
     - The executable file that will start an Ignite node process from `ignite-kubernetes/config/run.sh`
 
 Go to the folder and execute a command below to prepare the image:
```
Comment (Nikita-tech-writer) [inline modules/kubernetes/DEVNOTES.md:36]:
```suggestion
To prepare the image, navigate to the folder and execute the following command:
```

File: modules/kubernetes/DEVNOTES.md [hunk 12]
```diff
@@ -34,7 +34,7 @@ Create a folder for all the files to be placed in the Docker image and copy the
     - The executable file that will start an Ignite node process from `ignite-kubernetes/config/run.sh`
```
Comment (Nikita-tech-writer) [inline modules/kubernetes/DEVNOTES.md:34]:
```suggestion
    - The executable file that starts the Ignite node process from `ignite-kubernetes/config/run.sh`
```

File: modules/kubernetes/config/Dockerfile [hunk 13]
```diff
@@ -18,14 +18,17 @@
...
+ARG IGNITE_CFG_XML="example-kube.xml"
+
 # Set Apache Ignite version.
-ENV IGNITE_VERSION 2.0.0-SNAPSHOT
+ARG IGNITE_VERSION="2.0.0"
 
 # Set IGNITE_HOME variable.
 ENV IGNITE_HOME /opt/ignite/apache-ignite-${IGNITE_VERSION}-bin
 
 # Setting a path to an Apache Ignite configuration file. Used by run.sh script below.
```
Comment (Nikita-tech-writer) [inline modules/kubernetes/config/Dockerfile:30]:
```suggestion
# Set a path to the Apache Ignite configuration file. Use the run.sh script below:
```

File: modules/kubernetes/config/Dockerfile [hunk 14]
```diff
@@ -18,14 +18,17 @@
...
-ENV CONFIG_URI ${IGNITE_HOME}/config/example-kube.xml
+ENV CONFIG_URI ${IGNITE_HOME}/config/$IGNITE_CFG_XML
 
 # Make sure kubernetes lib is copied to 'libs' folder.
```
Comment (Nikita-tech-writer) [inline modules/kubernetes/config/Dockerfile:33]:
```suggestion
# Make sure the Kubernetes lib is copied to the 'libs' folder.
```

File: modules/kubernetes/config/Dockerfile [hunk 15]
```diff
@@ -51,7 +54,7 @@ COPY ./run.sh $IGNITE_HOME/
 RUN chmod +x $IGNITE_HOME/run.sh
 
 # Copying the configuration.
```
Comment (Nikita-tech-writer) [inline modules/kubernetes/config/Dockerfile:56]:
```suggestion
# Copy the configuration.
```

File: modules/kubernetes/config/Dockerfile [hunk 16]
```diff
@@ -51,7 +54,7 @@ COPY ./run.sh $IGNITE_HOME/
...
-COPY ./example-kube.xml $IGNITE_HOME/config
+COPY ./$IGNITE_CFG_XML $IGNITE_HOME/config
 
 # Starting an Apache Ignite node.
```
Comment (Nikita-tech-writer) [inline modules/kubernetes/config/Dockerfile:59]:
```suggestion
# Start an Apache Ignite node.
```

**Review by Nikita-tech-writer — COMMENTED** [review]
@timoninmaxim I've added some more minor comments

File: examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java (hunk 9 above)
Comment (timoninmaxim) [reply examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java:34]: Thin client does not "provide" access to Ignite, but do "have" the access to it within the cluster. I mean that there is no access to cluster if thin client isn't running within the cluster.

File: docs/_docs/thin-clients/java-thin-client.adoc (hunk 6 above)
Comment (timoninmaxim) [reply docs/_docs/thin-clients/java-thin-client.adoc:97]: @dmagda @Nikita-tech-writer  
To mention this only implementation is may be important to provide an existing example of the interface implementation.

Yes, your understanding is correct.

File: docs/_docs/installation/kubernetes/generic-configuration.adoc (hunk 4 above)
Comment (timoninmaxim) [reply docs/_docs/installation/kubernetes/generic-configuration.adoc:398]: I mean that list of addresses in ClientConfiguration must be in sync with list of addresses of IgniteConfiguration.

**Comment by timoninmaxim** [comment]
@Nikita-tech-writer I've commited most of you suggestions, thanks! I wrote a comment for the only non-commited suggestion.

Also I'm going to fix the comment @dmagda about docs with the `fetchServerAddresses` method. I will prepare a patch for it today, than ask @Nikita-tech-writer to review it.

File: examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java (hunk 9 above)
Comment (Nikita-tech-writer) [reply examples/src/main/java/org/apache/ignite/examples/client/ClientKubernetesPutGetExample.java:34]: OK, got it, thanks for the explanation!

**Review by Nikita-tech-writer — COMMENTED** [review]
@timoninmaxim I approve all the changes. 
@dmagda we need your word and approval.

File: docs/_docs/installation/kubernetes/generic-configuration.adoc (hunk 4 above)
Comment (timoninmaxim) [reply docs/_docs/installation/kubernetes/generic-configuration.adoc:398]: Agree, looks confusting. Remove this sentence at all.

File: docs/_docs/thin-clients/java-thin-client.adoc (hunk 6 above)
Comment (timoninmaxim) [reply docs/_docs/thin-clients/java-thin-client.adoc]: @Nikita-tech-writer I've updated the docs. Could you please check it?

Comment (Nikita-tech-writer) [reply docs/_docs/thin-clients/java-thin-client.adoc]: @timoninmaxim Everything looks good to me! I approve the PR.

**Review by Nikita-tech-writer — APPROVED** [review]

**Review by Mmuzaf — APPROVED** [review]

**Pull request merged**